Data/HashMap/Base.hs

```diff
@@ -751,12 +1083,131 @@ update f = alter (>>= f)
+"alterFinsert" forall (f :: Maybe a -> Identity (Maybe a)) x k m.
+  alterFWeird (coerce (Just x)) (coerce (Just x)) f k m =
+    Identity (insert k x m)
+
+-- This rule handles the case where 'alterF' is used to do 'insertWith'-like
+-- things. Whenever possible, GHC will get rid of the Maybe nonsense for us.
+-- We delay this rule to stage 1 so alterFinsert has a chance to fire.
+"alterFinsertWith" [1] forall (f :: Maybe a -> Identity (Maybe a)) x y k m.
+  alterFWeird (coerce (Just x)) (coerce (Just y)) f k m =
+    Identity (insertModifying x (\mold -> case runIdentity (f (Just mold)) of
+                                            Nothing -> bogus# (# #)
+                                            Just new -> (# new #)) k m)
+
+-- Handle the case where someone uses 'alterF' instead of 'adjust'. This
+-- rule is kind of picky; it will only work if the function doesn't
+-- do anything between case matching on the Maybe and producing a result.
+"alterFadjust" forall (f :: Maybe a -> Identity (Maybe a)) _y k m.
+  alterFWeird (coerce Nothing) (coerce (Just _y)) f k m =
+    Identity (adjust# (\x -> case runIdentity (f (Just x)) of
+                               Just x' -> (# x' #)
+                               Nothing -> bogus# (# #)) k m)
+
+
+-- Here we know that f doesn't use its argument at all.
+"alterFdelete" forall (f :: Maybe a -> Identity (Maybe a)) k m.
+  alterFWeird (coerce Nothing) (coerce Nothing) f k m = Identity (delete k m)
+
+-- The simple specialization to Const; in this case we can look up
+-- the key without caring what position it's in.
+"alterFlookup" forall _ign1 _ign2 (f :: Maybe a -> Const r (Maybe a)) k m.
+  alterFWeird _ign1 _ign2 f k m = k `seq` m `seq` Const (getConst (f (lookup k m)))
+ #-}
+
+alterFWeird
+       :: (Functor f, Eq k, Hashable k)
+       => f (Maybe v)
+       -> f (Maybe v)
+       -> (Maybe v -> f (Maybe v)) -> k -> HashMap k v -> f (HashMap k v)
+alterFWeird _ _ f k m = alterFEager f k m
+{-# INLINE [0] alterFWeird #-}
+
+alterFEager :: (Functor f, Eq k, Hashable k)
+       => (Maybe v -> f (Maybe v)) -> k -> HashMap k v -> f (HashMap k v)
+alterFEager f !k m = (<$> f mv) $ \fres ->
+  case fres of
+
+    ------------------------------
+    -- Delete the key from the map.
+    Nothing -> case lookupRes of
+
+      -- Key did not exist in the map to begin with, no-op
+      Absent -> m
+
+      -- Key did exist
+      Present _ collPos -> deleteKeyExists collPos h k m
+
+    ------------------------------
+    -- Update value
+    Just v' -> case lookupRes of
+
+      -- Key did not exist before, insert v' under a new key
+      Absent -> insertNewKey h k v' m
+
+      -- Key existed before, no hash collision
+      Present v collPos ->
+        if v `ptrEq` v'
+        -- If the value is identical, no-op
+        then m
+        -- If the value changed, update the value.
+        else insertKeyExists collPos h k v' m
+
+  where !h = hash k
+        !lookupRes = lookupRecordCollision h k m
```

Comment:
Any reason why this function doesn't have a version for `GLASGOW_HASKELL >= 802` that uses `lookupRecordCollision#` to completely bypass the `LookupRes` constructors? Same for the version in the Strict module.

Reply:
I need to check the Core (or you can--start practicing reading `-ddump-simpl`!) but I believe GHC likely unboxes that for us.

Reply:
I need to check the Core (or you can--start practicing reading `-ddump-simpl`!) but I believe GHC likely unboxes that for us.

Reply:
Learning to read Core is something that's been on my todo list. I'll take a look but it'll probably look like gibberish to me at this point :P

Reply:
It mostly reads like badly formatted Haskell with terrible variable names. Significant differences:

* Type and dictionary arguments are all explicit.

* `case` is *always* strict.

* Recursion is done explicitly with top-level `Rec` or local `letrec`.

* Strict constructors are treated as functions that call the real underlying constructors.

Reply:
For the first while, I recommend using `-dsuppress-all -dno-suppress-type-signatures` to cut down on the clutter. You can reintroduce extra info later.

Reply:
Oh yes, another difference: in the latest versions, you can read `join` and `joinrec` as `let` and `letrec` and ignore `jump`, but know that they're generally more efficient.

Reply:
Thanks for the info! By "case is always strict" do you mean that the value being `case`d on is always fully evaluated?

Reply:
It's always evaluated to WHNF. Also, the default case is not `_` at the bottom but `DEFAULT` on top, maybe with underscores (don't remember). And the forced value sometimes gets its own name. So Haskell

```haskell
f !x = g x
```

may turn into something like

```
f = \x -> case x of x' { DEFAULT -> g x' }
```

Another thing to note is that there is no infix. Haskell `x +# y` will look in Core like `+# x y`.

Reply:
I've verified that GHC elides the `LookupRes` nonsense as expected. I think it's typically better to limit the ugly hand-unboxing to where we really need it. The only potential exception is the unboxed unary tuple, which is really easy to deal with.